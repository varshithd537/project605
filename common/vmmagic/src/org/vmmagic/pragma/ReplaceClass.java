/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package org.vmmagic.pragma;

import java.lang.annotation.Retention;
import java.lang.annotation.Target;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.ElementType;
import org.vmmagic.Pragma;

/**
 * The ReplaceClass annotation indicates that the members of this class annotated with
 * <code>ReplaceMember</code> will replace members who have same name and same descriptor
 * in the targeting class. The <code>value()</code> of this annotation should be the
 * descriptor of targeting class.
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.TYPE)
@Pragma
public @interface ReplaceClass {

  /**
   * @return the name of the class to be replaced, e.g. {@code java.lang.System}
   */
  String className();
}
