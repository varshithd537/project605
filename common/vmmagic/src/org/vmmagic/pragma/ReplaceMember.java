/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package org.vmmagic.pragma;

import java.lang.annotation.Retention;
import java.lang.annotation.Target;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.ElementType;
import org.vmmagic.Pragma;

/**
 * The ReplaceMember annotation indicates that this member will replace the member
 * who has the same name, or has the name given by {@code value()}, and same
 * descriptor in the targeting class which is referred by the {@code value()} of
 * {@code ReplaceClass} annotation.
 */
@Retention(RetentionPolicy.RUNTIME)
@Target({ElementType.FIELD, ElementType.METHOD})
@Pragma
public @interface ReplaceMember {
  String value() default "";
  String descriptor() default "";
}
