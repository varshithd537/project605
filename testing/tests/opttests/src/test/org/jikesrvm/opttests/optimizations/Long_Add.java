/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package test.org.jikesrvm.opttests.optimizations;

import org.vmmagic.pragma.NoInline;

public class Long_Add {

  private static final long TwoToThePowerOf32 = 1L << 32;

  public static void main(String[] args) {
    Long_Add long_and = new Long_Add();
    long_and.main();
  }

  private void main() {
    System.out.println("--- Tests for long addition ---");
    System.out.println("--- Add zero ---");
    addZero(0L);
    addZero(-1L);
    addZero(1L);
    addZero(Long.MAX_VALUE);
    addZero(Long.MIN_VALUE);
    addZero(Integer.MAX_VALUE);
    addZero(Integer.MIN_VALUE);
    System.out.println("--- And own value  ---");
    addOwnValue(0L);
    addOwnValue(-1L);
    addOwnValue(1L);
    addOwnValue(Long.MAX_VALUE);
    addOwnValue(Long.MIN_VALUE);
    addOwnValue(Integer.MAX_VALUE);
    addOwnValue(Integer.MIN_VALUE);
    System.out.println("--- And negative value ---");
    addNegativeValue(0L);
    addNegativeValue(-1L);
    addNegativeValue(1L);
    addNegativeValue(Long.MAX_VALUE);
    addNegativeValue(Long.MIN_VALUE);
    addNegativeValue(Integer.MAX_VALUE);
    addNegativeValue(Integer.MIN_VALUE);
    System.out.println("--- Add one ---");
    addOne(0L);
    addOne(-1L);
    addOne(1L);
    addOne(Long.MAX_VALUE);
    addOne(Long.MIN_VALUE);
    addOne(Integer.MAX_VALUE);
    addOne(Integer.MIN_VALUE);
    addOne(2L);
    System.out.println("--- Add minus one ---");
    addMinusOne(0L);
    addMinusOne(-1L);
    addMinusOne(1L);
    addMinusOne(Long.MAX_VALUE);
    addMinusOne(Long.MIN_VALUE);
    addMinusOne(Integer.MAX_VALUE);
    addMinusOne(Integer.MIN_VALUE);
    addMinusOne(TwoToThePowerOf32);
    System.out.println("--- Add Long.MAX_VALUE ---");
    addLongMaxValue(0L);
    addLongMaxValue(-1L);
    addLongMaxValue(1L);
    addLongMaxValue(Long.MAX_VALUE);
    addLongMaxValue(Long.MIN_VALUE);
    addLongMaxValue(Integer.MAX_VALUE);
    addLongMaxValue(Integer.MIN_VALUE);
    addLongMaxValue(TwoToThePowerOf32);
    System.out.println("--- Add Long.MIN_VALUE ---");
    addLongMinValue(0L);
    addLongMinValue(-1L);
    addLongMinValue(1L);
    addLongMinValue(Long.MAX_VALUE);
    addLongMinValue(Long.MIN_VALUE);
    addLongMinValue(Integer.MAX_VALUE);
    addLongMinValue(Integer.MIN_VALUE);
    addLongMinValue(TwoToThePowerOf32);
  }

  @NoInline
  private static void addZero(long val) {
    long res = val + 0L;
    System.out.println(Long.toString(val) + " + 0L: " + Long.toString(res));
  }

  @NoInline
  private static void addOwnValue(long val) {
    long res = val + val;
    System.out.println(Long.toString(val) + " + " + Long.toString(val) + "L: " + Long.toString(res));
  }

  @NoInline
  private static void addNegativeValue(long val) {
    long res = val + -val;
    System.out.println(Long.toString(val) + " + -" + Long.toString(val) + "L: " + Long.toString(res));
  }

  @NoInline
  private static void addOne(long val) {
    long res = val + 1L;
    System.out.println(Long.toString(val) + " + 1L: " + Long.toString(res));
  }

  @NoInline
  private static void addMinusOne(long val) {
    long res = val + -1;
    System.out.println(Long.toString(val) + " + -1L: " + Long.toString(res));
  }

  @NoInline
  private static void addLongMaxValue(long val) {
    long res = val + Long.MAX_VALUE;
    System.out.println(Long.toString(val) + " + " + Long.MAX_VALUE + "L: " + Long.toString(res));
  }

  @NoInline
  private static void addLongMinValue(long val) {
    long res = val + Long.MIN_VALUE;
    System.out.println(Long.toString(val) + " + " + Long.MAX_VALUE + "L: " + Long.toString(res));

  }

}
