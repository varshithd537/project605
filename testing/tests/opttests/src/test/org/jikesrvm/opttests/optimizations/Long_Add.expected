--- Tests for long addition ---
--- Add zero ---
0 + 0L: 0
-1 + 0L: -1
1 + 0L: 1
9223372036854775807 + 0L: 9223372036854775807
-9223372036854775808 + 0L: -9223372036854775808
2147483647 + 0L: 2147483647
-2147483648 + 0L: -2147483648
--- And own value  ---
0 + 0L: 0
-1 + -1L: -2
1 + 1L: 2
9223372036854775807 + 9223372036854775807L: -2
-9223372036854775808 + -9223372036854775808L: 0
2147483647 + 2147483647L: 4294967294
-2147483648 + -2147483648L: -4294967296
--- And negative value ---
0 + -0L: 0
-1 + --1L: 0
1 + -1L: 0
9223372036854775807 + -9223372036854775807L: 0
-9223372036854775808 + --9223372036854775808L: 0
2147483647 + -2147483647L: 0
-2147483648 + --2147483648L: 0
--- Add one ---
0 + 1L: 1
-1 + 1L: 0
1 + 1L: 2
9223372036854775807 + 1L: -9223372036854775808
-9223372036854775808 + 1L: -9223372036854775807
2147483647 + 1L: 2147483648
-2147483648 + 1L: -2147483647
2 + 1L: 3
--- Add minus one ---
0 + -1L: -1
-1 + -1L: -2
1 + -1L: 0
9223372036854775807 + -1L: 9223372036854775806
-9223372036854775808 + -1L: 9223372036854775807
2147483647 + -1L: 2147483646
-2147483648 + -1L: -2147483649
4294967296 + -1L: 4294967295
--- Add Long.MAX_VALUE ---
0 + 9223372036854775807L: 9223372036854775807
-1 + 9223372036854775807L: 9223372036854775806
1 + 9223372036854775807L: -9223372036854775808
9223372036854775807 + 9223372036854775807L: -2
-9223372036854775808 + 9223372036854775807L: -1
2147483647 + 9223372036854775807L: -9223372034707292162
-2147483648 + 9223372036854775807L: 9223372034707292159
4294967296 + 9223372036854775807L: -9223372032559808513
--- Add Long.MIN_VALUE ---
0 + 9223372036854775807L: -9223372036854775808
-1 + 9223372036854775807L: 9223372036854775807
1 + 9223372036854775807L: -9223372036854775807
9223372036854775807 + 9223372036854775807L: -1
-9223372036854775808 + 9223372036854775807L: 0
2147483647 + 9223372036854775807L: -9223372034707292161
-2147483648 + 9223372036854775807L: 9223372034707292160
4294967296 + 9223372036854775807L: -9223372032559808512
